WordPress/gutenberg
# Block List: Display inserter button between blocks

Previously: #2890
Related: #2983

This pull request seeks to display a block inserter between blocks by hover or tab focus, enabling a user to easily insert between blocks. Originally this was introduced in #2890 and later reverted in #2929 pending a design review. This is a first step toward eventually dropping the header global inserter (#2755).

The changes here use the design reference from #2983:

![insert-between](https://user-images.githubusercontent.com/1204802/31434144-d50161c8-ae7b-11e7-9e99-07b596a0c00f.png)

As implemented:

![insert-between](https://user-images.githubusercontent.com/1779930/31499379-9944d9aa-af32-11e7-98be-dac15c4b318e.gif)

__Implementation notes:__

Since hovering in the specific area between blocks with a cursor can require a difficult amount of precision, I used a technique to provide a buffer for the cursor while the inserter is visible. A similar example of this sort of cursor intent buffer can be seen [in this article](http://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown) about Amazon.com's menu behavior. Consider the following visualization of the effect:

![hover-range](https://user-images.githubusercontent.com/1779930/31499596-3dcf03f6-af33-11e7-986f-2033a216d0b5.gif)

Changes include a straight revert of the revert in #2929, plus additional changes per design recommendations.

I chose to implement the sibling inserter as only rendering the underlying inserter component while it should be considered visible. This added some complexity to the implementation, but was in an effort to reduce the number of nodes on the page at a given time. I was (and still do) experience significant lag when inserting between on the demo content. It's not clear whether this is an issue specific to these changes, and further profiling is required.

__Testing instructions:__

Verify that you can insert between blocks, either by hovering between blocks or tabbing from a previous block, and that regressions do not otherwise occur with the global inserter or end-of-content inserters.

1. Navigate to Gutenberg > New Post
2. Note that no between-serters are available while no blocks are present
3. Insert a paragraph
4. Note that hovering before or after the new paragraph block reveals between-serters
   - Similarly for tabbing from paragraph

Some trickier behaviors to try:

- Pressing escape from the inserter should return focus to the inserter button and it should remain visible
- Moving the cursor outside the between-serter region while the inserter is open should not close the inserter
- Clicking outside the inserter without making a selection should cause the between-serter to become hidden
